src/rsa/mod.rs

```diff
@@ -0,0 +1,268 @@
+}
+
+impl ReadFrom for PublicKey {
+    fn read_from<R: std::io::Read>(input: &mut R) -> Result<Self> {
+        let buf_size = input.read_u16::<BigEndian>()?;
+        let n_size = input.read_u16::<BigEndian>()?;
+
+        let mut buf = vec![0u8; buf_size as usize];
+        input.read_exact(&mut buf)?;
+        let (n_buf, e_buf) = buf.split_at(n_size as usize);
+
+        let n = rsa::BigUint::from_bytes_be(n_buf);
+        let e = rsa::BigUint::from_bytes_be(e_buf);
+        let public_key = RsaPublicKey::new(n, e)?;
+        Ok(PublicKey(public_key))
+    }
+}
+
+impl Hash for PublicKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.hash(state)
+    }
+}
+
+impl PartialEq for PublicKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl PublicKey {
+    pub fn public_key_size(&self) -> usize {
+        self.0.n().to_bytes_be().len() + self.0.e().to_bytes_be().len() + 5
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{Keypair, TryFrom};
+    use crate::{Network, Sign, Verify};
+    use hex_literal::hex;
+    use rand::rngs::OsRng;
+
+    #[test]
+    fn sign_roundtrip() {
+        let keypair = Keypair::generate(Network::MainNet, &mut OsRng);
+        let signature = keypair.sign(b"hello world").expect("signature");
+        assert!(keypair
+            .public_key
+            .verify(b"hello world", &signature)
+            .is_ok())
+    }
+
+    #[test]
+    fn bytes_roundtrip() {
+        use rand::rngs::OsRng;
+        let keypair = Keypair::generate(Network::MainNet, &mut OsRng);
+        let bytes = keypair.to_vec();
+        assert_eq!(
+            keypair,
+            super::Keypair::try_from(&bytes[..]).expect("keypair")
+        );
+    }
+
+    #[test]
+    fn verify() {
+        // Test a msg signed and verified with a keypair generated with erlang crypto
+        // and compressed by hand.
+        const MSG: &[u8] = b"hello world";
+        const PUBKEY: &str = "1trSusebcQv2kJfLEUV1D4RQyHZyTfFkvFxWBUa1iv53eZKhyg1iDWGsWo89w8HzQBx3vzoeB85aDYK9w2oX1LdWdnrq5QL4M8iGDDacdp5FeSvXTwr6RB9Hv86qQSFT3ppdTSk6Jbe8eDK81NcNNrkhRXqfmH3CAHRCmrKwLcNBLzxo2a8hqQi1rsW8z9dJgWKMsx2cWoboaGgqrfsRC54WJuPWZwkRCmP7dHArxyWqibicaicBoq5yqW3QsTvxTXLHMUVXr59BQriu75QFiztCYiFjq13Qp6kVkFdXwZ5S2cSVZSsg9d1uB4eN3VK4wYefKFnR9qQT5S93CFFX9nXQx7wi5Z6MdAj1mmu6yZczCE";
+        let public_key: crate::PublicKey = PUBKEY.parse().expect("b58 public key");
+        const SIG: &[u8] =
+            &hex!("315050906cd1d58e056e6d9cd0311621a3ad04a60f4b778803c53da335be8239592420cd1910b91fee50fca6d030150356bf86bbe7066b5476da5016f988ec58c1d25fe1d50651de839c897d8e07eb2612e77f44abceef14d40a3358aa93498cd0361516ea29684f94001ea1800beae9cb5a701b371bcbd556b26a125f8fe4b7e9fd4395f5ad1bc80f48c0d40de56d14b2cd2fe109b4ba9cd26fe06dd79d2c236364448873b6a0aeccff87dba44db49b3fb2adb9274c09480e4ee9f281b92314c0fd375e862d349f32f2479775e60c98e11fefc6d99908238862640f356aaa755922dc7932433c4a1ab256c114255651f6b17b3076e6b06b0b79093a7409f03b");
```

Comment:
What exactly is this signature format, btw? I'll be asking the same question below, as well.

Reply:
all signatures in this library are DER encoded.

Reply:
I'm not able to decode this particular signature in an online DER encoder. Have some of the bytes been removed?

Reply:
not removed.. lemme figure this out. What's the link to the decoder?.. it's a standard call to the rust crypto rsa library..

Reply:
Disclaimer: this could easily be a PEBKAC error, but I've usually found this site to be pretty good: https://lapo.it/asn1js/


I'll also see if I can do it via OpenSSL.

Reply:
I guess I was wrong about the DER encoding... that's for PSS signatures which doesn't work. https://github.com/RustCrypto/RSA/blob/master/src/pkcs1v15/signature.rs#L16-L23